### ISSUE TITLE
Maybe config.EXIT_ON_ERROR should not exit immediately but throw an error

### ISSUE THREAD

**Issue by lamerman** [lamerman/shellpy#45]
Now if config.EXIT_ON_ERROR is set any returncode != 0 will make the script exit immediately. Maybe an exception should be thrown instead which, if not intercepted, is caught on the very top. When caught, a stacktrace may be printed and the same returncode returned in the handler.

**Comment by lamerman** [comment]
A core.NonZeroReturnCodeError is thrown on error in experimental mode or if config.EXIT_ON_ERROR is enabled.

**Comment by lamerman** [comment]
@gpalsingh @Jack614 @tildebyte @nikudalov guys I wanted to ask for your opinion as you used more or less shellpython.
In this task I started to raise a core.NonZeroReturnCodeError on every command that does not return 0. This is experimental feature and is not turned on by default. But it changes a lot the way errors are handled. Before you handled them as
```
result = `ls -l mydir
if result:
    `echo 1 > mydir/my.txt
```
But now it's completely different
```
result = `ls -l mydir
`echo 1 > mydir/my.txt
```
If `ls -l mydir` fails in the second example, an exception will be raised and the script will stop working printing out the failed line, returncode of `ls` and stderr of `ls`.

You also can handle those exceptions as usual in python
```
try:
  result = `ls -l mydir
except NonZeroReturnCodeError:
  `mkdir mydir

`echo 1 > mydir/my.txt
```

Rationale for this is that most likely people do not handle errors for all lines of shell code and there is not way to make them handle it like in Go (you are forced to handled error or explicitly say that you do not handle it). Thus it is easy to skip an error in shell and continue execution with it not even noticing the error.

To run scripts in experimental mode simply add -x parameter to shellpy like shellpy -x my.spy. If you use shellpython directly from python, then set config.EXIT_ON_ERROR to True, it will have the same behavior.

So, what do you think? Is it a better approach for error handling and shellpython in general? Should we go with it in the future or stay with the current approach when no exceptions are thrown?

**Comment by gpalsingh** [comment]
After some thought I think that it should be default that the script stops execution. People would most likely use the result of the command later (if not they would use bash instead of shellpy) which will throw an exception. Even if it didn't, it could potentially become a bug which could be very hard to remove.
Also the stack trace (as you know) prints extra information which needs to be fixed first (I can't work on it because I'm little busy these days).

**Comment by lamerman** [comment]
@gpalsingh I'm also thinking now about making it as default behaviour, right now it's experimental but very soon it should become default in my view. First I want to try it on real scripts if there are problems with this approach.
About the stacktrace I actually tried to make it but did not understand how could it be done. I had problems when I tried to remove all shellpy frames from the stack when you catch exception in the very same place in code that raises it and try to modify it there. It seems to be easier if you modify it on the top of stacktrace, not where it's thrown, but it all needs additional work and research...
